src/views/1_ApplicationComponent.tsx

```diff
@@ -9,33 +9,77 @@ import {saveWindowBounds} from "./ViewUtils";
 import {StatusBarComponent} from "./StatusBarComponent";
 import {PaneTree, Pane} from "../utils/PaneTree";
 import {SearchComponent} from "./SearchComponent";
+import {SplitDirection} from "../Enums";
 
 export class ApplicationComponent extends React.Component<{}, {}> {
     private tabs: Tab[] = [];
     private focusedTabIndex: number;
+    private electronWindow: Electron.BrowserWindow = remote.getCurrentWindow();
 
     constructor(props: {}) {
         super(props);
-        const electronWindow = remote.BrowserWindow.getAllWindows()[0];
 
         this.addTab(false);
 
-        electronWindow
-            .on("move", () => saveWindowBounds(electronWindow))
+        this.electronWindow
+            .on("move", () => saveWindowBounds(this.electronWindow))
             .on("resize", () => {
-                saveWindowBounds(electronWindow);
+                saveWindowBounds(this.electronWindow);
                 this.recalculateDimensions();
-            })
-            .webContents
+            });
+
+        this.electronWindow.webContents
             .on("devtools-opened", () => this.recalculateDimensions())
             .on("devtools-closed", () => this.recalculateDimensions());
 
-        ipcRenderer.on("change-working-directory", (_event: Electron.IpcRendererEvent, directory: string) =>
-            this.focusedTab.focusedPane.session.directory = directory,
-        );
+        ipcRenderer
+            // Tabs actions
+            .on("change-working-directory", (_event: Electron.IpcRendererEvent, directory: string) =>
```

Comment:
What about a class with two abstract methods `send()` and `receive()`. Each "message channel"  (e.g. change-working-directory) would be a subclass, where the programmer's task would be to make sure both methods have the same signature.

In the "emitting" process, one would call `send()`, which has the sole purpose of serializing its arguments using private methods and emit the ipc message. The "listening" process would subscribe to the channel by calling a method for that purpose on our class, causing our `receive()` method to be executed on the receipt of the message. The `receive()` call would be made internally by our class, which would take care of subscribing to the channel and deserializing the arguments.

Does that make any sense and would that even be helpful?

One problem i see here is that we'd have to bring other variables into the scope of the `receive()` method…

Here how the implementation for our "change-working-directory" example could look like:
```TypeScript
class ChangeWorkingDirectory extends RemoteCall {
	/**
	 * Only there to call `serializeAndEmit` with all its arguments.
	 */
	send(directory: string) {  // same signature
		this.serializeAndEmit({directory: directory});
	}
	/**
	 * Implement the functionality here.
	 */
	receive(directory: string) {  // same signature
		this.context.focusedTab.focusedPane.session.directory = directory;
		// context could be passed in on event subscription
	}
}
```